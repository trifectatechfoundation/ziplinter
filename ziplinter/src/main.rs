use rc_zip_sync::ReadZip;
use std::ops::Deref;

fn main() {
    let mut args = std::env::args();
    let _ = args.next();
    let path = args.next().unwrap();

    let file = std::fs::File::open(path).unwrap();
    let archive = file.read_zip().unwrap();

    eprintln!(
        "{}",
        serde_json::to_string_pretty(&archive.deref()).unwrap()
    )
}

#[cfg(test)]
mod test {
    use super::*;

    use insta::assert_json_snapshot;
    use std::{error::Error, path::Path};

    fn process_zip_file(zip_path: &Path) -> Result<serde_json::Value, Box<dyn Error>> {
        let file = std::fs::File::open(zip_path).unwrap();
        let archive = file.read_zip();

        Ok(serde_json::to_value(archive?.deref())?)
    }

    #[test]
    fn snapshot_zip_files() {
        let fixtures_dir = std::env::current_dir()
            .unwrap()
            .join("../testdata")
            .canonicalize()
            .unwrap();

        println!("fixtures_dir: {}", fixtures_dir.display());

        for entry in std::fs::read_dir(fixtures_dir).expect("Failed to read fixtures directory") {
            let entry = entry.expect("Failed to read entry");
            let path = entry.path();

            if path
                .extension()
                .is_some_and(|e| e.eq_ignore_ascii_case("zip"))
            {
                let Ok(result) = process_zip_file(&path) else {
                    continue;
                };
                println!(
                    "current file: {}",
                    Path::new(path.file_name().unwrap()).display()
                );
                assert_json_snapshot!(
                    format!("{}", Path::new(path.file_name().unwrap()).display()),
                    result
                );
            }
        }
    }
}
