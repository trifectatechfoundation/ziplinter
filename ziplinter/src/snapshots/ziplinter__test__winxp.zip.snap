---
source: ziplinter/src/main.rs
expression: result
---
{
  "comment": "",
  "contents": [
    {
      "central": {
        "comment": "",
        "compressed_size": 8,
        "crc32": 2098461837,
        "creator_version": {
          "host_system": "Mvs",
          "version": 20
        },
        "disk_nbr_start": 0,
        "external_attrs": 32,
        "extra": [],
        "flags": 0,
        "header_offset": 0,
        "internal_attrs": 1,
        "method": "Store",
        "mode": 0,
        "modified": "2011-12-08T10:04:24Z",
        "name": "hello",
        "reader_version": {
          "host_system": "MsDos",
          "version": 10
        },
        "uncompressed_size": 8
      },
      "local": {
        "accessed": null,
        "compressed_size": 8,
        "crc32": 2098461837,
        "created": null,
        "extra": [],
        "flags": 0,
        "gid": null,
        "header_offset": 0,
        "method": "Store",
        "method_specific": "None",
        "mode": 0,
        "modified": "2011-12-08T10:04:24Z",
        "name": "hello",
        "reader_version": {
          "host_system": "MsDos",
          "version": 10
        },
        "uid": null,
        "uncompressed_size": 8
      }
    },
    {
      "central": {
        "comment": "",
        "compressed_size": 6,
        "crc32": 2055117726,
        "creator_version": {
          "host_system": "Mvs",
          "version": 20
        },
        "disk_nbr_start": 0,
        "external_attrs": 32,
        "extra": [],
        "flags": 0,
        "header_offset": 43,
        "internal_attrs": 1,
        "method": "Store",
        "mode": 0,
        "modified": "2011-12-08T10:04:50Z",
        "name": "dir/bar",
        "reader_version": {
          "host_system": "MsDos",
          "version": 10
        },
        "uncompressed_size": 6
      },
      "local": {
        "accessed": null,
        "compressed_size": 6,
        "crc32": 2055117726,
        "created": null,
        "extra": [],
        "flags": 0,
        "gid": null,
        "header_offset": 0,
        "method": "Store",
        "method_specific": "None",
        "mode": 0,
        "modified": "2011-12-08T10:04:50Z",
        "name": "dir/bar",
        "reader_version": {
          "host_system": "MsDos",
          "version": 10
        },
        "uid": null,
        "uncompressed_size": 6
      }
    },
    {
      "central": {
        "comment": "",
        "compressed_size": 0,
        "crc32": 0,
        "creator_version": {
          "host_system": "Mvs",
          "version": 20
        },
        "disk_nbr_start": 0,
        "external_attrs": 16,
        "extra": [],
        "flags": 0,
        "header_offset": 86,
        "internal_attrs": 0,
        "method": "Store",
        "mode": 2147483648,
        "modified": "2011-12-08T10:08:06Z",
        "name": "dir/empty/",
        "reader_version": {
          "host_system": "MsDos",
          "version": 10
        },
        "uncompressed_size": 0
      },
      "local": {
        "accessed": null,
        "compressed_size": 0,
        "crc32": 0,
        "created": null,
        "extra": [],
        "flags": 0,
        "gid": null,
        "header_offset": 0,
        "method": "Store",
        "method_specific": "None",
        "mode": 2147483648,
        "modified": "2011-12-08T10:08:06Z",
        "name": "dir/empty/",
        "reader_version": {
          "host_system": "MsDos",
          "version": 10
        },
        "uid": null,
        "uncompressed_size": 0
      }
    },
    {
      "central": {
        "comment": "",
        "compressed_size": 12,
        "crc32": 3127775578,
        "creator_version": {
          "host_system": "Mvs",
          "version": 20
        },
        "disk_nbr_start": 0,
        "external_attrs": 33,
        "extra": [],
        "flags": 0,
        "header_offset": 126,
        "internal_attrs": 1,
        "method": "Store",
        "mode": 0,
        "modified": "2011-12-08T10:06:08Z",
        "name": "readonly",
        "reader_version": {
          "host_system": "MsDos",
          "version": 10
        },
        "uncompressed_size": 12
      },
      "local": {
        "accessed": null,
        "compressed_size": 12,
        "crc32": 3127775578,
        "created": null,
        "extra": [],
        "flags": 0,
        "gid": null,
        "header_offset": 0,
        "method": "Store",
        "method_specific": "None",
        "mode": 0,
        "modified": "2011-12-08T10:06:08Z",
        "name": "readonly",
        "reader_version": {
          "host_system": "MsDos",
          "version": 10
        },
        "uid": null,
        "uncompressed_size": 12
      }
    }
  ],
  "encoding": "Utf8",
  "eocd": {
    "dir": {
      "inner": {
        "dir_disk_nbr": 0,
        "dir_records_this_disk": 4,
        "directory_offset": 176,
        "directory_records": 4,
        "directory_size": 214,
        "disk_nbr": 0
      },
      "offset": 390
    },
    "dir64": null,
    "global_offset": 0
  },
  "parsed_ranges": [
    {
      "contains": "end of central directory record",
      "end": 412,
      "start": 390
    },
    {
      "contains": "central directory header",
      "end": 227,
      "filename": "hello",
      "start": 176
    },
    {
      "contains": "central directory header",
      "end": 280,
      "filename": "dir/bar",
      "start": 227
    },
    {
      "contains": "central directory header",
      "end": 336,
      "filename": "dir/empty/",
      "start": 331
    },
    {
      "contains": "central directory header",
      "end": 390,
      "filename": "readonly",
      "start": 491
    },
    {
      "contains": "local file header",
      "end": 35,
      "filename": "hello",
      "start": 0
    },
    {
      "contains": "file data",
      "end": 43,
      "filename": "hello",
      "start": 35
    },
    {
      "contains": "local file header",
      "end": 80,
      "filename": "dir/bar",
      "start": 43
    },
    {
      "contains": "file data",
      "end": 86,
      "filename": "dir/bar",
      "start": 80
    },
    {
      "contains": "local file header",
      "end": 126,
      "filename": "dir/empty/",
      "start": 86
    },
    {
      "contains": "file data",
      "end": 126,
      "filename": "dir/empty/",
      "start": 126
    },
    {
      "contains": "local file header",
      "end": 164,
      "filename": "readonly",
      "start": 126
    },
    {
      "contains": "file data",
      "end": 176,
      "filename": "readonly",
      "start": 164
    }
  ],
  "size": 412
}
